### PR TITLE
Ext ext msg explanation tooltip

### PR DESCRIPTION
**Pull request by martinraison**
@2is10 
(@stkem)

### CODE REVIEW

**Comment by 2is10** [comment]
Let me know when you’re ready, and I’ll take another look.

**Comment by martinraison** [comment]
I think it is ready know.

File: packrat/scripts/formatting.js [hunk 1]
```diff
@@ -216,13 +217,20 @@ var formatAuxData = (function () {
...
-      if (added.some(isMe)) {
-        return boldNamesOf(meInFront(added)) + ' were added by ' + nameOf(actor) + '.';
-      }
-      return nameOf(actor) + ' added ' + boldNamesOf(added) + '.';
+      var i = 0;
+      while (i < added.length && added[i].kind !== 'email') i++;
+      return render('html/keeper/message_aux_email', {
+        message: message,
+        hasEmail: i < added.length
+      });
```
Comment (2is10) [inline packrat/scripts/formatting.js]: I’d recommend we just return the message string, as before. The rendering of this template can be done at the call site. I’d rather not add a specific template dependency to this script.

File: packrat/scripts/thread.js [hunk 2]
```diff
@@ -81,6 +82,15 @@ panes.thread = function () {
     var $holder = $tall.find('.kifi-scroll-inner')
       .preventAncestorScroll()
       .handleLookClicks()
+      .hoverfu('.kifi-message-email-learn', function (configureHover) {
+        var link = this;
+        render('html/keeper/message_email_tooltip', function (html) {
+          configureHover(html, {
+            mustHoverFor: 9000000, click: 'toggle',
```
Comment (2is10) [inline packrat/scripts/thread.js]: `1e9` is the arbitrarily large number I’ve used elsewhere. It’s both compact and not large enough to risk causing [timer overflow](http://stackoverflow.com/questions/3468607/why-does-settimeout-break-for-large-millisecond-delay-values).

File: packrat/scripts/formatting.js (hunk 1 above)
Comment (2is10) [inline packrat/scripts/formatting.js]: In fact, could we just cut the content of the message_aux_email.html template and paste it into message_aux.html? (We’d need to pass a new `hasEmail` parameter to that template.)

**Comment by 2is10** [comment]
Feel free to merge when you’re happy with it. Nice work.

File: packrat/scripts/thread.js (hunk 2 above)
Comment (martinraison) [inline packrat/scripts/thread.js]: nice - didn't know about this

File: packrat/scripts/thread.js [hunk 3]
```diff
@@ -201,6 +211,11 @@ panes.thread = function () {
   function renderMessage(m) {
     m.formatMessage = formatMessage.full;
     m.formatAuxData = formatAuxData;
+    if (m.auxData && m.auxData.length >= 3 && m.auxData[0] === 'add_participants') {
+      var added = m.auxData[2], i = 0;
+      while (i < added.length && added[i].kind !== 'email') i++;
+      m.hasEmail = i < added.length;
```
Comment (2is10) [inline packrat/scripts/thread.js:217]: By the way, we have the luxury of assuming modern JS when coding in the extension. These three lines could also be written as:

```
m.hasEmail = m.auxData[2].some(function (o) {return o.kind === 'email'});
```

Comment (martinraison) [inline packrat/scripts/thread.js:217]: ah, much better. I just changed it - thanks!

**Pull request merged**